go/glome/glome.go

```diff
@@ -0,0 +1,199 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package glome implements GLOME protocol.
+package glome
+
+import (
+	"errors"
+	"fmt"
+	"io"
+
+	"crypto/curve25519"
+	"crypto/hmac"
+
+	"golang.org/x/crypto/curve25519"
+)
+
+const (
+	// PrivateKeySize is the size of a PrivateKey in bytes.
+	PrivateKeySize = 32
+	// PublicKeySize is the size of a PublicKey in bytes.
+	PublicKeySize = 32
+	// SharedSecretSize is the size of a SharedSecret in bytes.
+	SharedSecretSize = 32
+	// Maximum size allow for a Tag
+	MaxTagSize = 32
+	// Minimum size allow for a Tag
+	MinTagSize = 1
+)
+
+var (
+	ErrInvalidPublicKey = errors.New(
+		fmt.Sprintf("glome: invalid public key - byte slice len is not %d", PublicKeySize))
+	ErrInvalidPrivateKey = errors.New(
+		fmt.Sprintf("glome: invalid private key - byte slice len is not %d", PrivateKeySize))
+	ErrInvalidTagSize = errors.New(
+		fmt.Sprintf("glome: invalid tag size - MinPeerTagSize must be in range %d-%d",
+			MinTagSize, MaxTagSize))
+)
+
+// SharedSecret is the type of GLOME (and x25519) shared secret.
+type SharedSecret [SharedSecretSize]byte
+
+// PublicKey is the type of GLOME (and x25519) public Keys.
+type PublicKey [PublicKeySize]byte
+
+// Generate a PublicKey object from slice. Return ErrInvalidPublicKey
+// if slice's length is not PublicKeySize.
+func PublicKeyFromSlice(b []byte) (*PublicKey, error) {
```

Comment:
I'd tend to return byte slices as that, not as pointers. Especially as you don't want anyone to modify where the byte slice points and your byte slices are actually ephemeral (i.e. newly generated within the function).

Reply:
The only thing this function does is to take a slice and transform it into a pointer to PublicKey, so that there is a good interaction between both types. If this function returns a slice:

- The PublicKey type functionality would not be accessible from the object it returns.
- The only functionality of this function would be to check if the last slice had 32bytes. It would be better in that case to return only the existence of error.

Another question apart from the type that returns this function is if PublicKey should be of type [32]byte or []byte. I chose the first one so that the library could avoid dealing with errors in time of use, and only add this function for when it was convenient for the user to have a slice.